### ISSUE TITLE
Questions about keypoints

### ISSUE THREAD

**Issue by ChengkaiYang** [cocoon2wong/Vertical#10]
Dear @cocoon2wong :
       What a great job in trajectory prediction! But I still have some questions.  In AKL loss, we supervise coarse trajectory with keypoints. How do we get ground truth keypoints used for loss function? Are ground truth keypoints part of ground truth trajectory or ground truth keypoints are constructed artificially based on  ground truth of trajectory? If keypoints are constructed artificially based on ground truth, how can we guarantee it will bring more efficient or extra information to model to learn? (Imagine if we don't use such method, ground truth may be the only thing to supervise.If anything irrelevant or not such useful brought in model to supervise, does the performance of model drop?) If keypoints are part of ground truth, my question will be solved. Looking forward for your reply! Thank you very much!

**Comment by cocoon2wong** [comment]
Hi @ChengkaiYang !
Thank you for your questions!

These keypoints are part of groundtruth, we just need to determine their corresponding time indices.
In the average keypoint loss, we have
https://github.com/cocoon2wong/Vertical/blob/178866cf547150dc98d18817713e257aff7429f9/silverballers/agents/__baseAgent.py#L67-L74

Here, `self.p_index` is the indices of temporal keypoints. In our experiments, it will be set to `[3, 7, 11]` when forecasting $t_f = 12$ future points. Meanwhile, `len(self.p_index)` is the number of keypoints, i.e., the $N_{key}$ in our paper.

These different values were actually determined by ablation experiments as well. Please note that these keypoints are only used during training and are not needed during testing. Therefore, no additional information is introduced for training in our model.

I hope these answers can help you!

**Comment by cocoon2wong** [comment]
Hi @ChengkaiYang !

I noticed that you haven't replied to this question for a long time. If you have further questions, please feel free to ask! If you have no further questions, I will close this question later!

Thank you again for your question!
